File: src/python/modelmesh/connectors/rotation/stick_until_failure.py
"""Stick-until-failure rotation policy connector.

Bundles the three base rotation policy components (deactivation,
recovery, selection) into a single connector with a unified
configuration. Uses the current model until it fails, then rotates
to the next candidate.

Connector ID: ``modelmesh.stick-until-failure.v1``
"""
from __future__ import annotations

from dataclasses import dataclass

from modelmesh.cdk.base_rotation import (
    BaseDeactivationPolicy,
    BaseRecoveryPolicy,
    BaseRotationConfig,
    BaseSelectionStrategy,
)

__all__ = [
    "StickUntilFailureConfig",
    "StickUntilFailurePolicy",
]


@dataclass
class StickUntilFailureConfig(BaseRotationConfig):
    """Configuration for the stick-until-failure rotation policy.

    Inherits all settings from BaseRotationConfig, including
    ``failure_threshold``, ``cooldown_seconds``, ``error_rate_threshold``,
    ``request_limit``, ``token_limit``, ``budget_limit``, and
    ``model_priority``.
    """

    pass


class StickUntilFailurePolicy:
    """Stick-until-failure rotation policy bundle.

    Combines the three rotation policy components -- deactivation,
    recovery, and selection -- into a single connector with unified
    configuration. The policy keeps using the current model until it
    fails (exceeds the failure threshold or error rate), then deactivates
    it and selects the next best candidate.

    Connector ID: ``modelmesh.stick-until-failure.v1``

    The three component policies are:

    - **Deactivation**: Threshold-based. Deactivates a model when its
      consecutive failure count or error rate exceeds the configured
      thresholds, or when quota/budget limits are exceeded.
    - **Recovery**: Cooldown-based. Reactivates a standby model after
      the configured cooldown period has elapsed.
    - **Selection**: Stick-until-failure with priority support. Picks
      the first model from the priority list, or falls back to the
      candidate with the lowest error rate.

    Usage::

        policy = StickUntilFailurePolicy(StickUntilFailureConfig(
            failure_threshold=3,
            cooldown_seconds=60.0,
        ))
        should_deactivate = policy.deactivation.should_deactivate(state)
        should_recover = policy.recovery.should_recover(state)
        selected = policy.selection.select(candidates, request)
    """

    CONNECTOR_ID: str = "modelmesh.stick-until-failure.v1"

    def __init__(self, config: StickUntilFailureConfig | None = None) -> None:
        if config is None:
            config = StickUntilFailureConfig()
        self._config = config
        self._deactivation = BaseDeactivationPolicy(config)
        self._recovery = BaseRecoveryPolicy(config)
        self._selection = BaseSelectionStrategy(config)

    @property
    def deactivation(self) -> BaseDeactivationPolicy:
        """Return the deactivation policy component.

        Evaluates whether an active model should be moved to standby
        based on failure count, error rate, and quota limits.
        """
        return self._deactivation

    @property
    def recovery(self) -> BaseRecoveryPolicy:
        """Return the recovery policy component.

        Evaluates whether a standby model should be reactivated based
        on cooldown expiration.
        """
        return self._recovery

    @property
    def selection(self) -> BaseSelectionStrategy:
        """Return the selection strategy component.

        Chooses the best active model for a given request using
        priority list or lowest error rate.
        """
        return self._selection

File: src/typescript/src/connectors/rotation/stick-until-failure.ts
/**
 * Stick-until-failure rotation policy connector.
 *
 * Bundles deactivation, recovery, and selection policies into a single
 * connector with unified configuration.
 *
 * Connector ID: modelmesh.stick-until-failure.v1
 */

import { RuntimeEnvironment } from '../../interfaces/runtime';

export interface StickUntilFailureConfig {
  failureThreshold?: number;
  cooldownSeconds?: number;
  errorRateThreshold?: number;
  requestLimit?: number;
  tokenLimit?: number;
  budgetLimit?: number;
  modelPriority?: string[];
}

export class StickUntilFailurePolicy {
  static readonly CONNECTOR_ID = 'modelmesh.stick-until-failure.v1';
  static readonly RUNTIME = RuntimeEnvironment.UNIVERSAL;
  readonly config: Required<StickUntilFailureConfig>;

  constructor(config?: StickUntilFailureConfig) {
    this.config = {
      failureThreshold: config?.failureThreshold ?? 3,
      cooldownSeconds: config?.cooldownSeconds ?? 60,
      errorRateThreshold: config?.errorRateThreshold ?? 0.5,
      requestLimit: config?.requestLimit ?? 0,
      tokenLimit: config?.tokenLimit ?? 0,
      budgetLimit: config?.budgetLimit ?? 0,
      modelPriority: config?.modelPriority ?? [],
    };
  }

  /**
   * Determine whether a model should be deactivated based on its state.
   */
  shouldDeactivate(state: {
    consecutiveFailures: number;
    errorRate: number;
    totalRequests: number;
    totalTokens: number;
    totalCost: number;
  }): boolean {
    if (state.consecutiveFailures >= this.config.failureThreshold) return true;
    if (state.errorRate > this.config.errorRateThreshold && state.totalRequests > 0) return true;
    if (this.config.requestLimit > 0 && state.totalRequests >= this.config.requestLimit) return true;
    if (this.config.tokenLimit > 0 && state.totalTokens >= this.config.tokenLimit) return true;
    if (this.config.budgetLimit > 0 && state.totalCost >= this.config.budgetLimit) return true;
    return false;
  }

  /**
   * Determine whether a standby model should be recovered.
   */
  shouldRecover(state: {
    deactivatedAt?: Date;
  }): boolean {
    if (!state.deactivatedAt) return false;
    const elapsed = (Date.now() - state.deactivatedAt.getTime()) / 1000;
    return elapsed >= this.config.cooldownSeconds;
  }

  /**
   * Select the best candidate from active models.
   */
  select<T extends { modelId: string; errorRate: number }>(
    candidates: T[]
  ): T | null {
    if (candidates.length === 0) return null;

    // Try priority list first
    if (this.config.modelPriority.length > 0) {
      for (const priorityId of this.config.modelPriority) {
        const match = candidates.find((c) => c.modelId === priorityId);
        if (match) return match;
      }
    }

    // Fallback: lowest error rate
    return candidates.reduce((best, c) =>
      c.errorRate < best.errorRate ? c : best
    );
  }
}
